File: ChatApp-Electron/src/scripts/xss-util.js
/**
 * XSS防注入工具类 - 前端版本
 */
class XssUtil {
    /**
     * HTML转义字符映射表
     */
    static htmlEscapeMap = {
        '&': '&amp;',
        '<': '&lt;',
        '>': '&gt;',
        '"': '&quot;',
        "'": '&#x27;',
        '/': '&#x2F;'
    };

    /**
     * 清理HTML标签和特殊字符，防止XSS攻击
     * @param {string} input 原始输入
     * @returns {string} 清理后的安全字符串
     */
    static sanitize(input) {
        if (!input || typeof input !== 'string') {
            return input;
        }

        // HTML实体转义
        let sanitized = input.replace(/[&<>"'\/]/g, (char) => {
            return this.htmlEscapeMap[char];
        });

        // 过滤危险关键字
        sanitized = sanitized.replace(/javascript:/gi, '');
        sanitized = sanitized.replace(/on\w+\s*=/gi, ''); // onerror=, onclick=, onload= 等
        
        return sanitized;
    }

    /**
     * 显示时解码HTML实体（用于显示用户输入的内容）
     * @param {string} input 已转义的字符串
     * @returns {string} 解码后的字符串
     */
    static decode(input) {
        if (!input || typeof input !== 'string') {
            return input;
        }

        const textarea = document.createElement('textarea');
        textarea.innerHTML = input;
        return textarea.value;
    }

    /**
     * 移除所有HTML标签
     * @param {string} input 原始输入
     * @returns {string} 纯文本
     */
    static stripHtml(input) {
        if (!input || typeof input !== 'string') {
            return input;
        }
        return input.replace(/<[^>]*>/g, '');
    }

    /**
     * 验证输入长度
     * @param {string} input 输入字符串
     * @param {number} maxLength 最大长度
     * @returns {boolean} 是否有效
     */
    static validateLength(input, maxLength) {
        return input && input.length <= maxLength;
    }

    /**
     * 安全地设置元素的文本内容（自动转义）
     * @param {HTMLElement} element DOM元素
     * @param {string} text 文本内容
     */
    static setTextContent(element, text) {
        if (element && text !== null && text !== undefined) {
            element.textContent = text;
        }
    }

    /**
     * 安全地设置元素的HTML内容（已转义的内容）
     * @param {HTMLElement} element DOM元素
     * @param {string} html HTML内容（必须是已转义的）
     */
    static setHtmlContent(element, html) {
        if (element && html !== null && html !== undefined) {
            element.innerHTML = html;
        }
    }
}

// 导出供其他模块使用
if (typeof module !== 'undefined' && module.exports) {
    module.exports = XssUtil;
}

File: ChatApp-Backend/src/main/java/com/chatapp/util/XssUtil.java
package com.chatapp.util;

import org.springframework.web.util.HtmlUtils;

/**
 * XSS防注入工具类
 */
public class XssUtil {
    
    /**
     * 清理HTML标签和特殊字符，防止XSS攻击
     * @param input 原始输入
     * @return 清理后的安全字符串
     */
    public static String sanitize(String input) {
        if (input == null || input.isEmpty()) {
            return input;
        }
        
        // 使用Spring提供的HtmlUtils进行HTML转义
        String sanitized = HtmlUtils.htmlEscape(input);
        
        // 额外过滤一些危险字符
        sanitized = sanitized.replaceAll("javascript:", "");
        sanitized = sanitized.replaceAll("onerror=", "");
        sanitized = sanitized.replaceAll("onclick=", "");
        sanitized = sanitized.replaceAll("onload=", "");
        sanitized = sanitized.replaceAll("<script", "&lt;script");
        sanitized = sanitized.replaceAll("</script>", "&lt;/script&gt;");
        
        return sanitized;
    }
    
    /**
     * 批量清理字符串
     * @param inputs 原始输入数组
     * @return 清理后的安全字符串数组
     */
    public static String[] sanitize(String... inputs) {
        if (inputs == null) {
            return null;
        }
        
        String[] result = new String[inputs.length];
        for (int i = 0; i < inputs.length; i++) {
            result[i] = sanitize(inputs[i]);
        }
        return result;
    }
    
    /**
     * 验证输入长度
     * @param input 输入字符串
     * @param maxLength 最大长度
     * @return 是否有效
     */
    public static boolean validateLength(String input, int maxLength) {
        return input != null && input.length() <= maxLength;
    }
    
    /**
     * 移除所有HTML标签
     * @param input 原始输入
     * @return 纯文本
     */
    public static String stripHtml(String input) {
        if (input == null || input.isEmpty()) {
            return input;
        }
        return input.replaceAll("<[^>]*>", "");
    }
}
